.gitignore

```diff
@@ -7,3 +7,4 @@
 .venv
 htmlcov/
 snuba.egg-info/
+schema/
```

Comment:
git submodules! we use them in relay for various things and nobody complained yet. We drive updates through the makefile.

Reply:
I think whether to commit the schema or not is not the real issue. 
To me the schema should be imported like all other dependencies, which is as pypi packages, that way:
- we do not need to build custom automation to download the right version of the schema and to maintain it locally
- we do not have to reproduce this same automation for each python repo that needs to load the schema
- it is not up to snuba to decide how to turn the jsonschema into a type manageable by mypy. It is something we can generalize and reuse for all the python repos that need access to the schemas
- we still can track back which version we were relying on at a given point in time.
- we can still build automation to keep requirements.txt up to date if we wanted.

Now I cannot move to pypi packages here in this PR because that package does not exist. Creating it would be my next step. In the meantime I added the update in the post-merge hook so, at least, we are sure we keep the local schema up to date.

What do you think about this approach? If we are on board with the pypi dependency, then I would have no strong opinion regarding checking the schema in till the package is not ready.

Reply:
So I temporarily committed the schema, while I am working on the pypi package instead of using the submodule because it will be easier to revert and the automation to keep this in sync is simpler considering this is a throw away solution that will not be needed as soon as the schema is packaged properly.

Reply:
No strong preference on separately packaging the schema data. One way or another we're going to have to be able to download and build the schemas, and keep them up to date with a canonical source. I'm not sure it makes much of a difference if both of those steps happen in this package or are split across two packages until there is another project that actually makes use of that shared library. (I'm not sure who else needs this, maybe someone on the data side? Sentry is getting closer to Python 3 but I'm not sure their perspective on typing yet.)

Reply:
Regarding others needing this:
- the data project. Hoping it will still exist, it is basically a copy paste of our consumer and it is on python 3, they already expressed interest in using it.
- sentry is, as you mentioned, on its way. Even if they do not go all way into typing, I think there would be fairly easy case we could make to type check the code that makes changes to the event objects they receive from outside and that they provide to an external dependency (snuba).

More in general regarding downloading the schema and keeping it up to date, here is my reasoning for mypi:
building the schema: I want to try to build a pypi package that exports the built TypedDict, not just the schema. So no logic to build the schema is needed in snuba (and in the data project). This would be useful especially since we did not really settle down with 100% certainty on using this package for generating the type from jsonschema. There are still alternatives to try.
Not 100% sure this actually is feasible, but I will figure it out in the next days and in the worst case scenario we would just import the schema.

About downloading the schema:
I think here is where pypi is making our lives easier. It is the idiomatic way to import dependencies, so once the right version is in requirements.txt CI will run on that specific version (which means that as soon as we start running mypy in CI that will block any error), clients would be kept up to date with one line only of automation (that we should already have) to keep the installed packages up to date with regards of requirements.txt. It takes care of downloading the file from the right place only when needed without us writing anything. While committing the schema requires us to add automation to download it, validate it and ensure we do not commit changes to the schema by mistake.

About keeping the schema up to date:
it is true that automation is needed in all cases (unless we decide to bump versions manually), though, for mypi, we would just need to automatically create a PR to bump the version in requirements.txt (that can easily be tied to the release of a new version of the schema), which I like because the updates of the schemas would always be isolated from functional PRs and potentially we could have independent versions per schema (with multiple schemas).
For submodules we would have to automate the initial load of the project submodules and we would still need similar automation to generate PRs to pull the submodule and push it back to snuba. So there is the initial setup and then the periodic update.
When committing the schema we would also have to verify no changes are made and committed, which is not needed in the other cases. If we do not just follow master on the schema repo (depends on the release process that is not defined yet) we would still need the additional automation to bump the version name

Reply:
>we would still need similar automation to generate PRs to pull the submodule and push it back to snuba

fwiw I think there is virtually no difference between submodules and pypi in this regard, you can use dependabot for both. I guess the difference comes mostly down to release schedule and submodule initialization for local checkout